[jiakai-17] iP

jiakai-17 opened the pull request:
# Max - Your Personal Task Management Buddy 🚀

Max is ~your~ **_THE_** go-to personal task management buddy written in Java with a stunning JavaFX GUI! Say goodbye to the chaos and let Max make managing your tasks a breeze.

## Why Max?

- ✨ **Effortless Task Management**: Max simplifies task management with its intuitive _text-based interface_. No complex apps or steep learning curves - just chat with Max!

- ⚡️ **Lightning-Fast Speed**: Max is optimized for speed, with a fully _text-based input system_. No more fumbling around with your mouse or trackpad!

- 🌟 **User-Friendly Design**: Max is designed for everyone, regardless of technical expertise. You'll quickly become expert at managing your tasks!

## Features

- [x] Add tasks
- [x] Delete tasks
- [x] Mark tasks
- [x] Search tasks

## Requirements

- [x]  Windows/MacOS/Linux
- [x]  Java 11

## Quick Start

0. Download Java 11 from [here](https://www.oracle.com/sg/java/technologies/javase/jdk11-archive-downloads.html)
1. Download the latest jar file from the Releases tab.
2. Run with `java -jar Max.jar` from your terminal.

## Testimonial

> ⭐⭐⭐⭐⭐<br>
> Max is an absolute revelation for someone like me who despises GUI-based tools and swears by the efficiency of Vim, especially Neovim. This text-based wonder has completely won me over. It's like having a Vim-inspired task manager, and I couldn't be happier. <br>
> -L

## Developer?

This is just a fenced code block. Do not pay too much attention to this.

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

jiakai-17 committed:
feat: Level-0

a) Duke is now Max
b) Implement initial skeletal version

jiakai-17 committed:
feat: Level-1

Echoes user's commands. Exit with "bye" command.

jiakai-17 committed:
feat: Level-2

Max stores a list of texts entered by the user. Retrieve with "list" command.

jiakai-17 committed:
feat: Level-3

Able to mark/unmark Tasks as done

jiakai-17 committed:
feat: Level-4

Refactor code to accomodate more commands.

jiakai-17 committed:
feat: A-TextUiTesting

Add semi-automated testing

jiakai-17 committed:
feat: Level-5

Added error handling and update testcase

jiakai-17 committed:
feat: Level-6

Adds support to delete tasks. Also updated testcases.

jiakai-17 committed:
feat: A-Enums

Moved task logic to separate TaskList class

jiakai-17 committed:
style: add javadocs and checkstyle

jiakai-17 committed:
feat: Implement Level-7

* Add save/load functionality for tasks
* Update testcase to test task restore from save file
* Mark task fields as final

jiakai-17 committed:
Merge branch 'branch-Level-7'

jiakai-17 committed:
feat: Implement Level-8

* Add support for dates for Deadline and Events
* Update testcase to test date string parsing
* Remove try-catch for TaskList save/load

jiakai-17 committed:
Merge branch 'branch-Level-8'

jiakai-17 committed:
feat: Implement A-MoreOOP

* Add Ui class for input/output of messages
* Add Parser class to parse user input
* Add [ERROR] tag to error messages
* Update testcases

jiakai-17 committed:
Merge branch 'branch-A-MoreOOP'

jiakai-17 committed:
feat: A-Packages

* Refactor package extensions.* -> duke.*

jiakai-17 committed:
Merge branch 'branch-A-Packages'

jiakai-17 committed:
Merge branch 'add-gradle-support' into branch-A-Gradle

jiakai-17 committed:
feat: Add Gradle support

* Run and Build with Gradle

jiakai-17 committed:
Merge branch 'branch-A-Gradle'

jiakai-17 committed:
test: A-JUnit

* Add JUnit tests for Parser and TaskList
* Refactor TaskList to take in file name for saving

jiakai-17 committed:
Merge branch 'branch-A-JUnit'

jiakai-17 committed:
feat: Implement Level-9

* Find tasks by keyword
* Added JUnit tests

jiakai-17 committed:
docs: A-JavaDoc

* Add JavaDocs to methods

jiakai-17 committed:
style: A-CodingStandard

* Run Reformat Code
* Run AB-3 Checkstyle

jiakai-17 committed:
Merge branch 'branch-A-CodingStandard'

jiakai-17 committed:
Merge branch 'branch-A-JavaDoc'

jiakai-17 committed:
Merge branch 'branch-Level-9'

jiakai-17 committed:
refactor: Move Duke.java to duke package

* Update test scripts

jiakai-17 committed:
feat!: Implement Level-10

* Add a GUI for Duke using JavaFx.
* Remove CLI for Duke.
* Minor Javadoc style fixes.

* Note: You may see a warning message when running Duke.
This warning message can be ignored.
See: https://github.com/nus-cs2103-AY2324S1/forum/issues/72

BREAKING CHANGE: Methods return Strings instead of printing to stdout.

jiakai-17 committed:
Merge branch 'branch-Level-10'

jiakai-17 committed:
style: Implement A-CheckStyle

* Add CheckStyle config files from AB3.
* Add CheckStyle plugin to buid.gradle.

jiakai-17 committed:
Merge branch 'branch-A-CheckStyle'

sudarshan2401 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,229 @@
+        return "Hello! I'm Max!\nWhat can I do for you?";
+    }
+
+}
```
LGTM! I like your use of informative error messages and javadoc comments.

sudarshan2401 commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,75 @@
+public class Storage {
+
+    // The path where the data is saved.
+    private final Path saveFilePath;
```
Perhaps the name should be savedFilePath instead?

jiakai-17 replied:
Hmm, I originally meant it to be a save file, so `saveFilePath` refers to the path to a save file.

sudarshan2401 commented on src/main/java/duke/tasks/TaskList.java:
```diff
@@ -0,0 +1,306 @@
+            sb.append("\nYou have no tasks in your list.");
+        } else {
+            for (int i = 0; i < list.size(); i++) {
+                sb.append("\n").append((i + 1)).append(".").append(list.get(i));
```
Would it be possible to minimise the chaining of the appends?

jiakai-17 replied:
Okay, will take note in the next commit to increase readability. Thank you.

sudarshan2401 commented on src/main/java/duke/ui/DialogBox.java:
```diff
@@ -0,0 +1,99 @@
+        db.flip();
+        return db;
+    }
+}
```
LGTM! Good use of comments to explain process!

sudarshan2401 commented on src/test/java/duke/tasks/TaskListTest.java:
```diff
@@ -0,0 +1,219 @@
+    }
+
+    @Test
+    public void add_noDescription_exceptionThrown() {
```
Nit: possible to use less complex naming convention?

jiakai-17 replied:
I was following the format for naming of tests in the [Coding Standard](https://se-education.org/guides/conventions/java/intermediate.html):
>Underscores may be used in test method names using the following three part format `featureUnderTest_testScenario_expectedBehavior()`
>
>e.g. `sortList_emptyList_exceptionThrown()` `getMember_memberNotFount_nullReturned`

sudarshan2401 reviewed (commented):
Impressive coding standard! Just a few Nits!

Eola-Z commented on src/main/java/duke/exceptions/DukeException.java:
```diff
@@ -0,0 +1,17 @@
+ * A RuntimeException that is thrown by Duke.
+ * All exceptions thrown by Duke should be a subclass of DukeException.
+ */
+public class DukeException extends RuntimeException {
+
+    /**
+     * A RuntimeException that is thrown by Duke.
+     *
+     * @param message The message to be displayed when the exception is thrown.
+     */
+    public DukeException(String message) {
+        super(message);
+    }
+}
```
Good use of unique exception classes to handle errors!
I especially like that you also made subclasses of this class to specify the different types of error. 👍

Eola-Z commented on src/main/java/duke/tasks/TaskList.java:
```diff
@@ -0,0 +1,306 @@
+    }
+
+    // Returns "task" if there is only 1 task in the list, else returns "tasks".
+    private String taskOrTasks() {
```
'taskOrTasks' doesn't quite sound like a verb name. 
It's tricky but might there be a way to name this method as a verb?
So far the best I've come up with is 'printTaskCheckPlural' but that's not great either...

jiakai-17 replied:
Hmm, okay, I will try to rename it. Thank you for your feedback!

Eola-Z commented on src/main/java/duke/tasks/TaskList.java:
```diff
@@ -0,0 +1,306 @@
+     * @param description The description of the task to add.
+     * @return String message
+     */
+    public String add(TaskType taskType, String description) throws DukeIllegalArgumentException {
+        if (description.isBlank()) {
+            throw new DukeIllegalArgumentException("The description of a ToDo task cannot be blank.");
+        }
+        if (taskType != TaskType.TODO) {
+            throw new DukeIllegalArgumentException("Only ToDo tasks can be added with just a description.");
+        }
+        return this.add(taskType, description, "", "", "");
+    }
```
Maybe the methods might be clearer if it was specified which type of Task it accepted in the name?
I understand your approach though, so it's your call on whether this is something that needs to be changed.

Eola-Z commented on src/main/java/duke/tasks/TaskList.java:
```diff
@@ -0,0 +1,306 @@
+            "Noted. I've removed this task:%n%s%nNow you have %d %s in the list.";
+
+    // An ArrayList that stores the list of tasks.
+    protected final ArrayList<Task> list;
```
This wasn't the most intuitive name to me... 
Perhaps a more descriptive name than 'list' could improve readability?

jiakai-17 replied:
Got it, thank you for pointing it out!

Eola-Z reviewed (commented):
So I was supposed to review the naming in this PR but **wow**...

You did a *phenomenal* job with this! I could hardly find fault with the code.
Great work so far! ⭐ ⭐ ⭐

jiakai-17 committed:
refactor: Fix issues identified in peer review

The naming of certain variables are less than ideal.
Excessive `StringBuilder.append()` calls are used in some areas.
There is inconsistent use of `\n` and `%n` for Strings.

Refactoring these identified issues would make the code more readable,
and allow for cross-platform compatibility.

Let's,
* rename the variable `saveFilePath` to `pathToSaveFile` (#1)
* use `String.format()` to reduce the amount of append calls (#2)
* rename the method `taskOrTasks` to `formatTaskPlurality` (#3)
* rename the variable `list` in the `TaskList` class to `tasks` (#4)
* replace most occurrences of newlines `\n` with `String.format("%n")`
* update tests to reflect this change

These changes would address the identified issues and enhance the
overall quality of the codebase.

jiakai-17 committed:
feat: Implement A-Streams

The use of for loops to iterate over tasks is not very readable.

Stream-based operations provide a more declarative and expressive way
to handle collections, making the code more readable and maintainable.

Lets take use Streams to replace for loops when iterating over tasks in
TaskList.

jiakai-17 committed:
feat: Implement A-Assertions

* Add assert statements to validate assumptions at runtime

Assertions are disabled by default.
Run with the `-ea` flag to enable them.

jiakai-17 committed:
refactor: Implement A-CodeQuality

The codebase contains many magic literals used for formatting, as well
as long methods.

This makes the code hard to read and maintain.

Let's,
* extract these magic literals into constants
* extract long methods into smaller, shorter methods

jiakai-17 committed:
ci: Implement A-CI

* Add GitHub Actions workflow file
* Remove IO redirection tests

jiakai-17 committed:
Merge pull request #6 from jiakai-17/A-CI

ci: Implement A-CI

jiakai-17 committed:
Merge pull request #9 from jiakai-17/A-CodeQuality

refactor: Implement A-CodeQuality

jiakai-17 committed:
Merge branch 'master' into A-Streams

jiakai-17 committed:
Merge pull request #7 from jiakai-17/A-Streams

feat: Implement A-Streams

jiakai-17 committed:
Merge branch 'master' into A-Assertions

jiakai-17 committed:
Merge pull request #8 from jiakai-17/A-Assertions

feat: Implement A-Assertions

jiakai-17 committed:
feat: Implement C-Update

* Support easy editing of task details with the `edit` command

jiakai-17 committed:
Merge pull request #10 from jiakai-17/branch-C-Update

feat: Implement C-Update

jiakai-17 committed:
feat: Implement C-Help

* Add `help` command to show a list of commands and usage.

jiakai-17 committed:
Merge pull request #11 from jiakai-17/branch-C-Help

feat: Implement C-Help

jiakai-17 committed:
fix: Update parsing logic for edit command

The parsing logic for the edit command does not follow specification.

* Update logic of parsing edit command arguments
* Update testcases
* Update error messages

jiakai-17 committed:
refactor: Fix code quality issues in #12

* Shorten long methods
* Fix violations of coding standard
* Update testcases to test new commands

jiakai-17 committed:
ui: Compact UI

* Reduce default size of speech bubbles.
* Add VBox to ensure profile picture is at bottom left/right.

jiakai-17 committed:
docs: Add User Guide

* Update docs/README.md
* Upload docs/Ui.png
